hatch/src/assets/builder.rs

```diff
@@ -0,0 +1,123 @@
+use assets::{ TupKind, PlatformKind, Arch, ProjectAsset };
+use assets::config::Config;
+use assets::tuprules::Tuprules;
+use assets::test_tupfile::Tupfile as TestTupfile;
+use assets::tupfile::Tupfile;
+use assets::platform::{ Linux, MacOS, Windows };
+use assets::tupfile_ini::TupfileIni;
+use project::{ Project, ProjectKind };
+use std::path::PathBuf;
+use task;
+
+pub struct Builder {
+  assets: Vec<ProjectAsset>,
+}
+
+impl Builder {
+  pub fn new() -> Builder {
+    Builder { assets: Vec::new() }
+  }
+
+  pub fn from(project: &Project) -> Builder {
+    let mut builder = Builder::new();
+
+    builder.project(&TupKind::Config, project);
+    builder.project(&TupKind::TestTupfile, project);
+    builder.project(&TupKind::Tuprules, project);
+    builder.project(&TupKind::Tupfile, project);
+    builder.project(&TupKind::TupfileIni, project);
+
+    let platform_type = task::platform_type();
+    builder.platform(&platform_type, project);
+
+    builder
+  }
+
+  pub fn assets(&mut self) -> &Vec<ProjectAsset> {
+    &self.assets.as_ref()
+  }
+
+  pub fn project(&mut self, asset_kind: &TupKind, project: &Project) {
+    let asset = match *asset_kind {
+      TupKind::Config => self.config(project),
+      TupKind::TestTupfile => self.test_tupfile(project),
+      TupKind::Tuprules => self.tuprules(project),
+      TupKind::Tupfile => self.tupfile(project),
+      TupKind::TupfileIni => self.tupfile_ini(project),
+    };
+
+    self.assets.push(asset);
+  }
+
+  pub fn platform(&mut self, asset_kind: &PlatformKind, project: &Project) {
+    let asset = match *asset_kind {
+      PlatformKind::Linux => self.linux(project),
+      PlatformKind::MacOS => self.macos(project),
+      PlatformKind::Windows => self.windows(project)
+    };
+
+    self.assets.push(asset);
+  }
+
+  pub fn config(&self, project: &Project) -> ProjectAsset {
+    let project_path = project.path();
+    let contents = Config::new(project.name(), project.kind()).to_string();
+
+    ProjectAsset::new(project_path.to_path_buf(), Config::name(), contents)
+  }
+
+  pub fn test_tupfile(&self, project: &Project) -> ProjectAsset {
+    let asset_path = project.path().join("test");
+    let contents = TestTupfile::new().to_string();
+
+    ProjectAsset::new(asset_path, TestTupfile::name(), contents)
+  }
+
+  pub fn tuprules(&self, project: &Project) -> ProjectAsset {
+    let project_path = project.path();
+    let project_kind = project.kind();
+    let contents = match *project_kind {
+      ProjectKind::Library(ref lib_type) =>  {
```

Comment:
Can this be done without { and } ?

Reply:
In theory, but the main reason I did it that way was due to line length.